### PR TITLE
Fix memory store wait

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

The memory store wait does not currently respect max num objects returned. This breaks application code (e.g., rllib) that uses wait.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -419,30 +419,38 @@ Status CoreWorker::Wait(const std::vector<ObjectID> &ids, int num_objects,
         plasma_store_provider_->Wait(plasma_object_ids, num_objects, /*timeout_ms=*/0,
                                      worker_context_.GetCurrentTaskID(), &ready));
   }
+  RAY_CHECK(static_cast<int>(ready.size()) <= num_objects);
   if (memory_object_ids.size() > 0) {
     // TODO(ekl) for memory objects that are ErrorType::OBJECT_IN_PLASMA, we should
     // consider waiting on them in plasma as well to ensure they are local.
-    RAY_RETURN_NOT_OK(memory_store_provider_.Wait(
-        memory_object_ids, std::max(0, static_cast<int>(ready.size()) - num_objects),
+    RAY_RETURN_NOT_OK(memory_store_provider_->Wait(
```
[the remark this answers is not in the document]
Comment (edoakes) [reply src/ray/core_worker/core_worker.cc]: Then we can get rid of the `std::max` too

File: python/ray/actor.py [hunk 2]
```diff
@@ -389,7 +390,7 @@ def _remote(self,
         if kwargs is None:
             kwargs = {}
         if is_direct_call is None:
-            is_direct_call = False
+            is_direct_call = bool(os.environ.get("RAY_FORCE_DIRECT"))
```
Comment (zhijunfu) [inline python/ray/actor.py]: when would this env variable `RAY_FORCE_DIRECT` be set?

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc [hunk 3]
```diff
@@ -193,22 +197,19 @@ Status CoreWorkerMemoryStore::Get(const std::vector<ObjectID> &object_ids,
...
     }
 
     // Return if all the objects are obtained.
-    if (remaining_ids.empty()) {
-      return Status::OK();
-    }
-
-    if (object_ids.size() - remaining_ids.size() >= static_cast<size_t>(num_objects)) {
-      // Already get enough objects.
+    if (remaining_ids.empty() || count >= num_objects) {
```
Comment (zhijunfu) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc]:
```suggestion
    if (count == num_objects) {
```

from the above logic `count` cannot be larger than `num_objects`, and also it seems like it's not necessary to check if `remaining_ids.empty()`?

File: src/ray/core_worker/store_provider/memory_store/memory_store.cc [hunk 4]
```diff
@@ -87,6 +87,9 @@ void GetRequest::Wait() {
 
 void GetRequest::Set(const ObjectID &object_id, std::shared_ptr<RayObject> object) {
   std::unique_lock<std::mutex> lock(mutex_);
+  if (is_ready_) {
+    return;  // We have already hit the number of objects to return limit.
```
Comment (zhijunfu) [inline src/ray/core_worker/store_provider/memory_store/memory_store.cc]: is this necessary?

File: src/ray/core_worker/core_worker.cc [hunk 5]
```diff
@@ -419,30 +419,34 @@ Status CoreWorker::Wait(const std::vector<ObjectID> &ids, int num_objects,
         plasma_store_provider_->Wait(plasma_object_ids, num_objects, /*timeout_ms=*/0,
                                      worker_context_.GetCurrentTaskID(), &ready));
   }
-  if (memory_object_ids.size() > 0) {
+  RAY_CHECK(static_cast<int>(ready.size()) <= num_objects);
+  if (memory_object_ids.size() > 0 && static_cast<int>(ready.size()) < num_objects) {
```
Comment (edoakes) [inline src/ray/core_worker/core_worker.cc]: nit: please reverse the order of these conditions to match the two if statements below

Comment (edoakes) [reply src/ray/core_worker/core_worker.cc]: (or reverse both of those)

**Review by edoakes — APPROVED** [review]

**Pull request merged**